### ISSUE TITLE
add scale_pos_weight parameter for catboost [not really a bug]

### ISSUE THREAD

**Issue by gbiele** [mlr-org/mlr3extralearners#229]
### Description

First, thanks for working on the mlr3 eco system!
I find this extremely useful.

According to the catboost documentation one can use one of these parameters to deal with unbalanced data:

- _scale_pos_weight_
- _class_weights_
- _auto_class_weights_

It looks as if only the latter two are exposed in `mlr_learners_classif.catboost`, see e.g. [here](https://rdrr.io/github/mlr-org/mlr3extralearners/man/mlr_learners_classif.catboost.html).

However, _class_weights_ & _auto_class_weights_ can only be used with custom metrics.
I would like to stick with already implemented metrics. Would it be possible to use _scale_pos_weight_ in combination with existing metrics, and is it possible to add that parameter to `mlr_learners_classif.catboost`?

### Reproducible example

```
library(mlr3)
library(mlr3tuning)
library(mlr3extralearners)
learner = lrn("classif.catboost")
learner$param_set$values = list(
  scale_pos_weight = .1
  )
```

**Comment by sebffischer** [comment]
Hey, thank you for your interest in mlr3 and for raising this issue!

I am not sure why this parameter is missing and will  look into this tomorrrow (and also check whether other parameters are missing). 

If you need a solution today, it seems that both xgboost and lightgbm have the `scale_pos_weight` parameter.

**Comment by sebffischer** [comment]
If you read the documentation of `help(catboost.train)` it does not list the `scale_pos_weight` parameter, so my guess - without checking in detail  - is that it is not available. Why does `class_weights` not work in your case exactly?

Copying from the help page of `catboost.train()`, the docs say: 

            • class_weights

              Classes weights. The values are used as multipliers for
              the object weights.

              For example, for 3 class classification you could use:

              ‘c(0.85, 1.2, 1)’

              Default value:

              None (the weight for all classes is set to 1)

**Comment by sebffischer** [comment]
And no. Sometimes adding the line you suggested might be enought, but one always has to check that the parameter is correctly passed down to the training function in the definition of `.train()`

**Comment by sebffischer** [comment]
Ok i think there is some confusion here: 
I meant that when you want to add a parameter to a learner (like hypothetically `scale_pos_weight`, you need to ensure that it is passed correctly to the training function of the upstream package (in this case `catboost.train()`. I.e. the code starting from here: 
https://github.com/mlr-org/mlr3extralearners/blob/652ffeff512381cba66c10f25a1db0f2b7b11e70/R/learner_catboost_classif_catboost.R#L200

Currently the mlr3 interface to catboost does not allow for custom loss functions. 

Is it possible that you set the loss to cross-entropy? Can you try it with the default logloss?

**Comment by gbiele** [comment]
Thanks for your example, I am doing something wrong, but I don't understand what.

Here is my approach:

```
learner = lrn("classif.catboost")
learner$predict_type = "prob"

learner$param_set$values = list(
  learning_rate = to_tune(lower = 1e-03, upper = .5, logscale = TRUE),
  depth = to_tune(lower = 1, upper = 10),
  iterations = to_tune(lower = 1, upper = 2000),
  l2_leaf_reg = to_tune(lower = 1e-03, upper = 10, logscale = TRUE),
  class_weights = c(1/prop_positive, 1)
)

learner = po("scale") %>>% po("learner", learner)
learner = GraphLearner$new(learner)

task = as_task_classif(
  m.dt,
  task_type = "classif",
  target = "ADHD.NPR",
  positive = "1",
  id = "pred.ADHD")

future::plan(list("multisession", "sequential"))

# https://mlr3book.mlr-org.com/optimization.html#nested-resampling
set.seed(1235)
rr = tune_nested(
  method = "random_search",
  task = task,
  learner = learner,
  measure = msr("classif.auc"), 
  inner_resampling = rsmp("cv", folds = 3),
  outer_resampling = rsmp("repeated_cv", folds = 4, repeats = 5),
  term_evals = 5
)
```

and here is the resulting erros message:

> Error in catboost::catboost.train(learn_pool, NULL, pars) : 
> catboost/private/libs/options/catboost_options.cpp:604: class weights takes effect only with Logloss, MultiClass, MultiClassOneVsAll and user-defined loss functions
> This happened PipeOp classif.catboost's $train()

**Comment by gbiele** [comment]
OK, got it.

I was wrong to set the class weights with 
```
learner$param_set$values = list(
  ...
  class_weights = c(1/prop_positive, 1)
)
```

when I set the class weights when creating the learner as you did, it works.

**Comment by sebffischer** [comment]
Aaah I was confused for a second as well. 

Compare the following: 

``` r
library(mlr3extralearners)

l = lrn("classif.catboost")
l$param_set$values
#> $loss_function_twoclass
#> [1] "Logloss"
#> 
#> $loss_function_multiclass
#> [1] "MultiClass"
#> 
#> $logging_level
#> [1] "Silent"
#> 
#> $thread_count
#> [1] 1
#> 
#> $allow_writing_files
#> [1] FALSE
#> 
#> $save_snapshot
#> [1] FALSE
l$param_set$values = list(class_weights = c(1, 2))
l$param_set$values
#> $class_weights
#> [1] 1 2

l = lrn("classif.catboost", class_weights = c(1, 2))
l$param_set$values
#> $loss_function_twoclass
#> [1] "Logloss"
#> 
#> $loss_function_multiclass
#> [1] "MultiClass"
#> 
#> $logging_level
#> [1] "Silent"
#> 
#> $thread_count
#> [1] 1
#> 
#> $allow_writing_files
#> [1] FALSE
#> 
#> $save_snapshot
#> [1] FALSE
#> 
#> $class_weights
#> [1] 1 2

# recently introduced in mlr3:
l = lrn("classif.catboost")
l$set_values(class_weights = c(1, 2))
#> Error in eval(expr, envir, enclos): attempt to apply non-function
l$param_set$values
#> $loss_function_twoclass
#> [1] "Logloss"
#> 
#> $loss_function_multiclass
#> [1] "MultiClass"
#> 
#> $logging_level
#> [1] "Silent"
#> 
#> $thread_count
#> [1] 1
#> 
#> $allow_writing_files
#> [1] FALSE
#> 
#> $save_snapshot
#> [1] FALSE
```

<sup>Created on 2022-08-25 by the [reprex package](https://reprex.tidyverse.org) (v2.0.1)</sup>

**Comment by gbiele** [comment]
Thanks for a clarification!
A quick off topic, but still mlr3 question:

when I try to do hyperparameter tuning with the method "mbo" for a catboost model I get this error: 

> Error: Element with key 'regr.km' not found in DictionaryLearner! Did you mean 'regr.ksvm' / 'regr.lmer' / 'regr.M5Rules'?

Should I post this at `bbtok` or `mlr3tuing`?

**Comment by sebffischer** [comment]
Here: https://github.com/mlr-org/mlr3mbo 
(if you manually set the surrogate learner I believe you can circumvent this bug)

**Comment by sebffischer** [comment]
You can checkout that branch, I think it should work
https://github.com/mlr-org/mlr3mbo/tree/defaults

**Comment by gbiele** [comment]
Thanks a lot !
I got it working now!
